tls/s2n_post_handshake.c

```diff
@@ -167,12 +167,42 @@ S2N_RESULT s2n_post_handshake_recv(struct s2n_connection *conn)
     return S2N_RESULT_OK;
 }
 
+S2N_RESULT s2n_post_handshake_write_records(struct s2n_connection *conn, s2n_blocked_status *blocked)
+{
+    struct s2n_stuffer *message = &conn->handshake.io;
+
+    uint32_t remaining = 0;
+    while ((remaining = s2n_stuffer_data_available(message)) > 0) {
+        if (s2n_stuffer_data_available(&conn->out)) {
+            RESULT_GUARD_POSIX(s2n_flush(conn, blocked));
+        }
+
+        uint16_t max_payload_size = 0;
+        RESULT_GUARD(s2n_record_max_write_payload_size(conn, &max_payload_size));
+
+        struct s2n_blob fragment = { 0 };
+        uint32_t fragment_size = MIN(remaining, max_payload_size);
+        uint8_t *fragment_data = s2n_stuffer_raw_read(message, fragment_size);
+        RESULT_ENSURE_REF(fragment_data);
+        RESULT_GUARD_POSIX(s2n_blob_init(&fragment, fragment_data, fragment_size));
+
+        RESULT_GUARD(s2n_record_write(conn, TLS_HANDSHAKE, &fragment));
+        RESULT_GUARD_POSIX(s2n_flush(conn, blocked));
+    }
+
+    RESULT_GUARD_POSIX(s2n_stuffer_wipe(message));
+    return S2N_RESULT_OK;
+}
+
 int s2n_post_handshake_send(struct s2n_connection *conn, s2n_blocked_status *blocked)
 {
     POSIX_ENSURE_REF(conn);
 
+    POSIX_GUARD_RESULT(s2n_post_handshake_write_records(conn, blocked));
+
     POSIX_GUARD(s2n_key_update_send(conn, blocked));
     POSIX_GUARD_RESULT(s2n_tls13_server_nst_send(conn, blocked));
 
+    POSIX_GUARD(s2n_stuffer_resize(&conn->handshake.io, 0));
```

Comment:
Is this a safety feature to ensure the buffer is gone? 

In `s2n_negotiate_impl` the same buffer is free'd after a call to `s2n_post_handshake_send`. There is one other call in `s2n_sendv_with_offset_impl` that doesn't free the buffer after the call. But the user can (is expect to) call `s2n_connection_free_handshake` which also free this same buffer.

Reply:
It's not a safety feature. It's how s2n_post_handshake_send cleans up after itself.

s2n_connection_free_handshake is entirely optional and we can't assume the application will call it. Even if the application does call it, it is intended to be called once, not after every call to s2n_send. Any call to s2n_send can trigger post-handshake messages, which reallocate handshake.io.

I'd call [the resize in s2n_negotiate_impl](https://github.com/aws/s2n-tls/blob/main/tls/s2n_handshake_io.c#L1666-L1670) a safety feature. Technically it's redundant and we could remove it. However, I'm a little reluctant to. s2n_negotiate uses handshake.io, so should clean up handshake.io. s2n_post_handshake_send also uses handshake.io, so should clean up handshake.io.

For example, what if we decide that reusing handshake.io for post-handshake messages isn't worth it, and we switch to using a dedicated post_handshake.out stuffer instead? Obviously s2n_post_handshake_send would stop cleaning up handshake.io, since it wouldn't be using it anymore. So then nobody would be cleaning up handshake.io, even though s2n_negotiate still uses it.